File: cost_calc.py
import csv, io, math
from datetime import date, timedelta

TODAY = date(2026, 6, 24)
N = 7
CUTOFF = TODAY - timedelta(days=N)
PRIOR_CUTOFF = TODAY - timedelta(days=2*N)

RATES = {
    'claude-opus-4-7':          {'input':15.00,'output':75.00,'cache_read':1.50,'cache_write':18.75},
    'claude-sonnet-4-6':        {'input':3.00, 'output':15.00,'cache_read':0.30,'cache_write':3.75},
    'claude-haiku-4-5-20251001':{'input':0.80, 'output':4.00, 'cache_read':0.08,'cache_write':1.00},
}

data = open('/home/runner/work/aeon/aeon/memory/token-usage.csv').read()
reader = csv.DictReader(io.StringIO(data))

current_rows = []
prior_rows = []
malformed = 0
unknown_models = {}

for row in reader:
    try:
        d = date.fromisoformat(row['date'].strip())
        skill = row['skill'].strip()
        model = row['model'].strip()
        inp = int(row['input_tokens'])
        out = int(row['output_tokens'])
        cr  = int(row['cache_read'])
        cw  = int(row['cache_creation'])
    except Exception:
        malformed += 1
        continue

    if model not in RATES:
        unknown_models[model] = unknown_models.get(model, 0) + inp + out
        rates = RATES['claude-opus-4-7']
    else:
        rates = RATES[model]

    cost = (inp/1e6*rates['input'] + out/1e6*rates['output'] +
            cr/1e6*rates['cache_read'] + cw/1e6*rates['cache_write'])

    r = dict(date=d, skill=skill, model=model,
             input=inp, output=out, cache_read=cr, cache_write=cw, cost=cost)

    if d >= CUTOFF:
        current_rows.append(r)
    elif d >= PRIOR_CUTOFF:
        prior_rows.append(r)

print(f"Current window rows: {len(current_rows)}")
print(f"Prior window rows: {len(prior_rows)}")
print(f"Malformed: {malformed}")
print(f"Unknown models: {unknown_models}")

total_cost = sum(r['cost'] for r in current_rows)

def row_breakdown(rows):
    ti = to = tcr = tcw = 0.0
    for r in rows:
        rates = RATES.get(r['model'], RATES['claude-opus-4-7'])
        ti  += r['input']/1e6*rates['input']
        to  += r['output']/1e6*rates['output']
        tcr += r['cache_read']/1e6*rates['cache_read']
        tcw += r['cache_write']/1e6*rates['cache_write']
    return ti, to, tcr, tcw

ti, to, tcr, tcw = row_breakdown(current_rows)
total_runs = len(current_rows)

print(f"\nCURRENT {CUTOFF} to {TODAY}")
print(f"Total: ${total_cost:.4f}  runs={total_runs}")
print(f"  Input:${ti:.4f} Output:${to:.4f} CacheRead:${tcr:.4f} CacheWrite:${tcw:.4f}")

prior_cost = sum(r['cost'] for r in prior_rows)
print(f"\nPRIOR {PRIOR_CUTOFF} to {CUTOFF}: ${prior_cost:.4f}  runs={len(prior_rows)}")
if prior_cost > 0:
    wow = (total_cost - prior_cost) / prior_cost * 100
    print(f"WoW: {wow:+.1f}%")

# Per-skill
skill_data = {}
for r in current_rows:
    s = r['skill']
    if s not in skill_data:
        skill_data[s] = {'runs':0,'tokens':0,'cost':0.0,'costs':[],'rows':[]}
    skill_data[s]['runs'] += 1
    skill_data[s]['tokens'] += r['input']+r['output']+r['cache_read']+r['cache_write']
    skill_data[s]['cost'] += r['cost']
    skill_data[s]['costs'].append(r['cost'])
    skill_data[s]['rows'].append(r)

print("\nTOP 10 BY COST:")
top10 = sorted(skill_data.items(), key=lambda x: -x[1]['cost'])[:10]
for sk, v in top10:
    avg = v['cost']/v['runs'] if v['runs'] else 0
    print(f"  {sk}: runs={v['runs']} tok={v['tokens']:,} cost=${v['cost']:.4f} avg=${avg:.4f}")

model_data = {}
for r in current_rows:
    m = r['model']
    if m not in model_data:
        model_data[m] = {'runs':0,'tokens':0,'cost':0.0}
    model_data[m]['runs'] += 1
    model_data[m]['tokens'] += r['input']+r['output']+r['cache_read']+r['cache_write']
    model_data[m]['cost'] += r['cost']

print("\nPER MODEL:")
for m, v in sorted(model_data.items(), key=lambda x: -x[1]['cost']):
    print(f"  {m}: runs={v['runs']} tok={v['tokens']:,} cost=${v['cost']:.4f}")

# Anomaly detection
print("\nANOMALIES:")
sm_data = {}
for r in current_rows:
    key = (r['skill'], r['model'])
    if key not in sm_data:
        sm_data[key] = []
    sm_data[key].append(r)

anomaly_count = 0
for (sk, m), rows in sm_data.items():
    if len(rows) < 3:
        continue
    costs = [r['cost'] for r in rows]
    mu = sum(costs)/len(costs)
    variance = sum((c-mu)**2 for c in costs)/len(costs)
    sigma = math.sqrt(variance)
    for r in rows:
        if r['cost'] > mu + 2*sigma and r['cost'] > 0.10:
            anomaly_count += 1
            print(f"  {sk}/{m} {r['date']} cost=${r['cost']:.4f} mu=${mu:.4f} 2sig=${mu+2*sigma:.4f}")
            print(f"    inp={r['input']:,} out={r['output']:,} cw={r['cache_write']:,}")

if anomaly_count == 0:
    print("  None.")
print(f"Total anomalies: {anomaly_count}")

# WoW skill spikes
prior_skill = {}
for r in prior_rows:
    s = r['skill']
    if s not in prior_skill:
        prior_skill[s] = 0.0
    prior_skill[s] += r['cost']

print("\nSKILL WOW SPIKES (>=2x, prior>=$0.25):")
for sk, v in skill_data.items():
    if sk in prior_skill and prior_skill[sk] >= 0.25:
        ratio = v['cost'] / prior_skill[sk]
        if ratio >= 2.0:
            print(f"  {sk}: cur=${v['cost']:.4f} pri=${prior_skill[sk]:.4f} {ratio:.1f}x")

# Forecast
daily_avg = total_cost / N
monthly = daily_avg * 30
print(f"\nFORECAST: daily=${daily_avg:.4f} monthly=${monthly:.4f}")

# Optimization: model downgrade candidates
print("\nOPTIMIZATION CANDIDATES (downgrade):")
for sk, v in skill_data.items():
    rows = [r for r in current_rows if r['skill'] == sk and r['model'] == 'claude-opus-4-7']
    if not rows:
        continue
    avg_cost = v['cost'] / v['runs']
    if avg_cost <= 0.25:
        continue
    total_inp = sum(r['input'] for r in rows)
    total_out = sum(r['output'] for r in rows)
    if total_inp == 0:
        continue
    ratio = total_out / total_inp if total_inp > 0 else 999
    if ratio < 0.3:
        opus_inp_cost = total_inp/1e6*15.0
        opus_out_cost = total_out/1e6*75.0
        son_inp_cost = total_inp/1e6*3.0
        son_out_cost = total_out/1e6*15.0
        opus_total = opus_inp_cost + opus_out_cost
        son_total = son_inp_cost + son_out_cost
        if opus_total > 0:
            savings_pct = (opus_total - son_total) / opus_total
            est_weekly = v['cost'] * savings_pct
            print(f"  {sk}: avg_cost=${avg_cost:.4f} out/inp={ratio:.3f} est_savings=${est_weekly:.4f}/wk")

# Cache underuse check (direct: cache_read / (cache_read+input) < 0.2, avg cost > $0.10)
print("\nCACHE UNDERUSE:")
for sk, v in skill_data.items():
    total_inp = sum(r['input'] for r in v['rows'])
    total_cr = sum(r['cache_read'] for r in v['rows'])
    avg_cost = v['cost'] / v['runs']
    if avg_cost <= 0.10:
        continue
    denom = total_cr + total_inp
    if denom == 0:
        continue
    ratio = total_cr / denom
    if ratio < 0.2:
        print(f"  {sk}: cache_ratio={ratio:.3f} avg_cost=${avg_cost:.4f}")

File: .cost_calc.js
const fs = require('fs');

const CUTOFF = "2026-08-04";
const PRIOR_CUTOFF = "2026-07-28";
const N = 7;

const RATES = {
  "claude-opus-4-7":           {input: 15.00, output: 75.00, cache_read: 1.50, cache_write: 18.75},
  "claude-sonnet-4-6":         {input: 3.00,  output: 15.00, cache_read: 0.30, cache_write: 3.75},
  "claude-haiku-4-5-20251001": {input: 0.80,  output: 4.00,  cache_read: 0.08, cache_write: 1.00},
};

const lines = fs.readFileSync('memory/token-usage.csv', 'utf8').trim().split('\n');
const header = lines[0].split(',');

const rows_current = [], rows_prior = [];
let malformed = 0;
const unknown_models = new Set();

for (let i = 1; i < lines.length; i++) {
  const parts = lines[i].split(',');
  if (parts.length < 7) { malformed++; continue; }
  const [date, skill, model, inp_s, out_s, cr_s, cw_s] = parts;
  const inp = parseInt(inp_s), out = parseInt(out_s), cr = parseInt(cr_s), cw = parseInt(cw_s);
  if (isNaN(inp)||isNaN(out)||isNaN(cr)||isNaN(cw)) { malformed++; continue; }

  let rates = RATES[model];
  if (!rates) { unknown_models.add(model); rates = RATES["claude-opus-4-7"]; }

  const cost = inp/1e6*rates.input + out/1e6*rates.output + cr/1e6*rates.cache_read + cw/1e6*rates.cache_write;
  const ic   = inp/1e6*rates.input;
  const oc   = out/1e6*rates.output;
  const crc  = cr/1e6*rates.cache_read;
  const cwc  = cw/1e6*rates.cache_write;

  const rec = {date, skill, model, inp, out, cr, cw, cost, ic, oc, crc, cwc};
  if (date >= CUTOFF) rows_current.push(rec);
  else if (date >= PRIOR_CUTOFF) rows_prior.push(rec);
}

function total(rows) {
  return rows.reduce((a,r)=>a+r.cost,0);
}
function totalIC(rows) { return rows.reduce((a,r)=>a+r.ic,0); }
function totalOC(rows) { return rows.reduce((a,r)=>a+r.oc,0); }
function totalCRC(rows){ return rows.reduce((a,r)=>a+r.crc,0); }
function totalCWC(rows){ return rows.reduce((a,r)=>a+r.cwc,0); }

const cur = total(rows_current);
const pri = total(rows_prior);
const wow = pri > 0 ? (cur - pri)/pri*100 : null;

console.log(`TOTAL=${cur.toFixed(4)}`);
console.log(`INPUT=${totalIC(rows_current).toFixed(4)}`);
console.log(`OUTPUT=${totalOC(rows_current).toFixed(4)}`);
console.log(`CACHE_READ=${totalCRC(rows_current).toFixed(4)}`);
console.log(`CACHE_WRITE=${totalCWC(rows_current).toFixed(4)}`);
console.log(`PRIOR=${pri.toFixed(4)}`);
console.log(`WOW=${wow !== null ? wow.toFixed(1) : 'none'}`);
console.log(`RUNS=${rows_current.length}`);

// per-skill
const skillMap = {};
for (const r of rows_current) {
  if (!skillMap[r.skill]) skillMap[r.skill] = {runs:0, tokens:0, cost:0, costArr:[], rows:[]};
  skillMap[r.skill].runs++;
  skillMap[r.skill].tokens += r.inp+r.out+r.cr+r.cw;
  skillMap[r.skill].cost += r.cost;
  skillMap[r.skill].costArr.push(r.cost);
  skillMap[r.skill].rows.push(r);
}

const top10 = Object.entries(skillMap).sort((a,b)=>b[1].cost-a[1].cost).slice(0,10);
console.log('TOP10_START');
for (const [s,d] of top10) {
  const avg = d.cost/d.runs;
  console.log(`${s}|${d.runs}|${d.tokens}|${d.cost.toFixed(4)}|${avg.toFixed(4)}`);
}
console.log('TOP10_END');

// per-model
const modelMap = {};
for (const r of rows_current) {
  if (!modelMap[r.model]) modelMap[r.model] = {runs:0, tokens:0, cost:0};
  modelMap[r.model].runs++;
  modelMap[r.model].tokens += r.inp+r.out+r.cr+r.cw;
  modelMap[r.model].cost += r.cost;
}
console.log('MODELS_START');
for (const [m,d] of Object.entries(modelMap).sort((a,b)=>b[1].cost-a[1].cost)) {
  console.log(`${m}|${d.runs}|${d.tokens}|${d.cost.toFixed(4)}`);
}
console.log('MODELS_END');

// anomalies (per run, >= 3 runs in (skill,model) group)
const smMap = {};
for (const r of rows_current) {
  const k = r.skill+'|'+r.model;
  if (!smMap[k]) smMap[k] = [];
  smMap[k].push(r);
}

const anomalies = [];
for (const [k, group] of Object.entries(smMap)) {
  if (group.length < 3) continue;
  const costs = group.map(r=>r.cost);
  const mu = costs.reduce((a,b)=>a+b,0)/costs.length;
  const sigma = Math.sqrt(costs.map(c=>(c-mu)**2).reduce((a,b)=>a+b,0)/costs.length);
  for (const r of group) {
    if (r.cost > mu + 2*sigma && r.cost > 0.10) {
      anomalies.push({...r, mu, sigma});
    }
  }
}

// skill doubles vs prior
const priorSkill = {};
for (const r of rows_prior) {
  if (!priorSkill[r.skill]) priorSkill[r.skill] = 0;
  priorSkill[r.skill] += r.cost;
}

console.log('ANOMALIES_START');
for (const a of anomalies) {
  console.log(`${a.skill}|${a.model}|${a.date}|${a.cost.toFixed(4)}|${a.mu.toFixed(4)}|${a.sigma.toFixed(4)}|${a.inp}|${a.out}|${a.cw}`);
}
console.log('ANOMALIES_END');

console.log('DOUBLES_START');
for (const [skill,d] of Object.entries(skillMap)) {
  const prior = priorSkill[skill]||0;
  if (prior >= 0.25 && d.cost >= 2*prior) {
    console.log(`${skill}|${d.cost.toFixed(4)}|${prior.toFixed(4)}`);
  }
}
console.log('DOUBLES_END');

const daily = cur/N;
const monthly = daily*30;
console.log(`DAILY=${daily.toFixed(4)}`);
console.log(`MONTHLY=${monthly.toFixed(4)}`);

// optimization: model downgrade candidates
console.log('OPT_START');
const downgrade_candidates = [];
for (const [skill, d] of Object.entries(skillMap)) {
  const opusRows = d.rows.filter(r=>r.model==='claude-opus-4-7');
  if (!opusRows.length) continue;
  const avgCost = d.cost/d.runs;
  if (avgCost <= 0.25) continue;
  const ratios = opusRows.filter(r=>r.inp>0).map(r=>r.out/r.inp);
  if (!ratios.length) continue;
  const sorted = [...ratios].sort((a,b)=>a-b);
  const medRatio = sorted[Math.floor(sorted.length/2)];
  if (medRatio < 0.3) {
    const savings = d.cost * (1 - 18/90);
    downgrade_candidates.push({skill, avgCost, medRatio, savings, cost: d.cost});
  }
}
downgrade_candidates.sort((a,b)=>b.savings-a.savings);
for (const c of downgrade_candidates.slice(0,3)) {
  console.log(`DOWNGRADE|${c.skill}|${c.cost.toFixed(4)}|${c.avgCost.toFixed(4)}|${c.medRatio.toFixed(3)}|${c.savings.toFixed(4)}`);
}

// cache underuse
const cache_candidates = [];
for (const [skill, d] of Object.entries(skillMap)) {
  const avgCost = d.cost/d.runs;
  if (avgCost <= 0.10) continue;
  const totalCR = d.rows.reduce((a,r)=>a+r.cr,0);
  const totalInp = d.rows.reduce((a,r)=>a+r.inp,0);
  if (totalCR+totalInp === 0) continue;
  const ratio = totalCR/(totalCR+totalInp);
  if (ratio < 0.2) {
    cache_candidates.push({skill, ratio, avgCost, cost: d.cost});
  }
}
cache_candidates.sort((a,b)=>b.cost-a.cost);
for (const c of cache_candidates.slice(0,3)) {
  console.log(`CACHE|${c.skill}|${c.cost.toFixed(4)}|${c.avgCost.toFixed(4)}|${c.ratio.toFixed(3)}`);
}
console.log('OPT_END');

console.log(`UNKNOWN=${[...unknown_models].join(',') || 'none'}`);
console.log(`MALFORMED=${malformed}`);
